File: training-ground/src/training_ground/features.py
from __future__ import annotations

import math
from typing import Any, Mapping

import numpy as np

FEATURE_SIZE = 206


def _encode_count(n: int, out: np.ndarray, offset: int) -> None:
    out[offset] = 1.0 if n >= 1 else 0.0
    out[offset + 1] = 1.0 if n >= 2 else 0.0
    out[offset + 2] = 1.0 if n >= 3 else 0.0
    out[offset + 3] = (n - 3) / 2.0 if n > 3 else 0.0


def featurize(position: Mapping[str, Any]) -> np.ndarray:
    """Position JSON → length-206 float32 STM vector ([docs/domain/features.md])."""
    points = position["points"]
    on_roll = position["onRoll"]
    if on_roll not in ("p1", "p2"):
        raise ValueError(f"onRoll must be p1 or p2, got {on_roll!r}")

    vec = np.zeros(FEATURE_SIZE, dtype=np.float32)
    stm = on_roll
    opp = "p2" if on_roll == "p1" else "p1"

    for k in range(24):
        if on_roll == "p1":
            json_idx = k
            raw = points[json_idx]
            stm_n = max(int(raw), 0)
            opp_n = max(-int(raw), 0)
        else:
            json_idx = 23 - k
            raw = points[json_idx]
            stm_n = max(-int(raw), 0)
            opp_n = max(int(raw), 0)
        _encode_count(stm_n, vec, k * 4)
        _encode_count(opp_n, vec, 96 + k * 4)

    bar = position["bar"]
    off = position["off"]
    vec[192] = bar[stm] / 2.0
    vec[193] = bar[opp] / 2.0
    vec[194] = off[stm] / 15.0
    vec[195] = off[opp] / 15.0

    cube = position["cube"]
    vec[196] = math.log2(cube["value"]) / 6.0
    owner = cube["owner"]
    vec[197] = 1.0 if owner == "centered" else 0.0
    vec[198] = 1.0 if owner == stm else 0.0
    vec[199] = 1.0 if owner == opp else 0.0
    may = cube["mayDouble"]
    vec[200] = 1.0 if may[stm] else 0.0
    vec[201] = 1.0 if may[opp] else 0.0
    return vec

File: ts-core/src/features.ts
import type { Position } from "./types";

export const FEATURE_SIZE = 206;

function encodeCount(n: number, out: Float32Array, offset: number): void {
  out[offset] = n >= 1 ? 1 : 0;
  out[offset + 1] = n >= 2 ? 1 : 0;
  out[offset + 2] = n >= 3 ? 1 : 0;
  out[offset + 3] = n > 3 ? (n - 3) / 2 : 0;
}

/** Position JSON → length-206 float32 STM vector (docs/domain/features.md). */
export function featurize(position: Position): Float32Array {
  const onRoll = position.onRoll;
  if (onRoll !== "p1" && onRoll !== "p2") {
    throw new Error(`onRoll must be p1 or p2, got ${String(onRoll)}`);
  }
  const vec = new Float32Array(FEATURE_SIZE);
  const stm = onRoll;
  const opp = onRoll === "p1" ? "p2" : "p1";
  const points = position.points;

  for (let k = 0; k < 24; k++) {
    let stmN: number;
    let oppN: number;
    if (onRoll === "p1") {
      const raw = points[k]!;
      stmN = Math.max(raw, 0);
      oppN = Math.max(-raw, 0);
    } else {
      const raw = points[23 - k]!;
      stmN = Math.max(-raw, 0);
      oppN = Math.max(raw, 0);
    }
    encodeCount(stmN, vec, k * 4);
    encodeCount(oppN, vec, 96 + k * 4);
  }

  vec[192] = position.bar[stm] / 2;
  vec[193] = position.bar[opp] / 2;
  vec[194] = position.off[stm] / 15;
  vec[195] = position.off[opp] / 15;

  const cube = position.cube;
  vec[196] = Math.log2(cube.value) / 6;
  const owner = cube.owner;
  vec[197] = owner === "centered" ? 1 : 0;
  vec[198] = owner === stm ? 1 : 0;
  vec[199] = owner === opp ? 1 : 0;
  vec[200] = cube.mayDouble[stm] ? 1 : 0;
  vec[201] = cube.mayDouble[opp] ? 1 : 0;
  return vec;
}
